### PR TITLE
`qemu-utils`: reduce build closure size

### PR DESCRIPTION
**Pull request by roberth**
## Description of changes

Like #251046, but an extra step smaller by omitting many build dependencies from the `qemu` input.

- qemu-utils: Rename parameter to qemu
  
  Allows overriding without prior knowledge of which qemu package attribute name is the default.

- qemu-utils: Use smaller qemu_test as input

  1823 -> 990 derivations in the build closure.

  This saves rebuilds and makes qemu-utils less susceptible to build errors.

  qemu_test and qemu_utils both need network- and disk-related functionality, so they should be a good match.

**Alternative I rejected**

I have considered an alternate solution to my problem. The NixOS test framework would tell `make-disk-image.nix` to use its `virtualisation.qemu`, so that `qemu-utils` isn't needed at all, but this would cause images to be rebuilt when switching to the interactive driver, which uses a fancier `qemu` with graphical dependencies, etc. I consider this to be too magical and a place for heisenbugs to hide. [Code](https://github.com/NixOS/nixpkgs/compare/master...hercules-ci:nixpkgs:nixos-test-disk-image-reuse-qemu) for posterity.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
  - Tested with `nixosTests.gitlab` after setting `virtualisation.writableStore = true;` so that `qemu-utils` is used.
  - `nix-store -qR $(nix-instantiate -A nixosTests.gitlab) | grep gtk` (`gtk-doc` is expected; `gtk+` would not be)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by alyssais** [comment]
I'm not sure about this — it has those positive effects now, but one of the things that makes qemu_test qemu_test is that it's patched in ways that are specific to the NixOS test runner.  I don't think the one patch we currently have would affect qemu-utils, but I'm concerned that in future that might not be the case.

**Comment by cole-h** [comment]
Is the patch what makes qemu_test small? If not, I wonder if it would make sense to add a qemu_super_minimal or something that has the same small closure size without NixOS specific patch(es), and build qemu_test on top of that...?

**Comment by alyssais** [comment]
Then we end up with yet another QEMU build.  I'm not sure that's an improvement.  (There's also the secret QEMU build defined in lib (eww) used for emulation.)

**Comment by roberth** [comment]
Found a better way. An empty list of architectures can be specified, in which case it produces mostly just the tools, in about 80 seconds.

<details>

Without extraction (ie just overrides; internal package without attribute (for now?))

```console
$ ls result/
bin/     include/ libexec/ share/

$ ls result/bin/
elf2dmp  qemu-edid  qemu-img  qemu-io  qemu-kvm  qemu-nbd  qemu-pr-helper  qemu-storage-daemon

$ nix-store -q --references result/
/nix/store/ld03l52xq2ssn4x0g5asypsxqls40497-glibc-2.37-8
/nix/store/ig0kkzw4n2pws12dj7szjm71f1a43if6-zlib-1.3
/nix/store/1c6x8n47mq5rcivsp72b0dyqgjghi6yx-glib-2.76.4
/nix/store/3b3mzw23bi8ga9jg8lh2pp0j61qnib77-liburing-2.4
/nix/store/5p7jmbpmq7clb3whbi6kyp50gmg54l2d-zstd-1.5.5
/nix/store/axyr8b6dmv24vxric98248hrb3nqnf7l-libaio-0.3.113
/nix/store/bhmynyjwzc2r6iqf7fhc3yzjcv3paiwa-curl-8.3.0
/nix/store/vawph0vsv3zwvknk6rhml51lx33lcspd-libcap-ng-0.8.3
/nix/store/646d9mid09n8h2zl8196l7912hziw92n-qemu-8.1.2

$ nix-store -q --size result
41949776
```

With extraction (ie `qemu-utils` proper)

```console
$ ls result/
bin  share

$ ls result/bin/
qemu-img  qemu-io  qemu-nbd

$ nix-store -q --size result
7361824

$ nix-store -q --references result/
/nix/store/ld03l52xq2ssn4x0g5asypsxqls40497-glibc-2.37-8
/nix/store/ig0kkzw4n2pws12dj7szjm71f1a43if6-zlib-1.3
/nix/store/1c6x8n47mq5rcivsp72b0dyqgjghi6yx-glib-2.76.4
/nix/store/3b3mzw23bi8ga9jg8lh2pp0j61qnib77-liburing-2.4
/nix/store/5p7jmbpmq7clb3whbi6kyp50gmg54l2d-zstd-1.5.5
/nix/store/axyr8b6dmv24vxric98248hrb3nqnf7l-libaio-0.3.113
/nix/store/bhmynyjwzc2r6iqf7fhc3yzjcv3paiwa-curl-8.3.0
```

</details>

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -34933,7 +34933,12 @@ with pkgs;
     inherit (darwin) sigtool;
   };
 
-  qemu-utils = callPackage ../applications/virtualization/qemu/utils.nix { };
+  qemu-utils = callPackage ../applications/virtualization/qemu/utils.nix {
+    qemu = qemu.override (o: {
+      toolsOnly = true;
+      hostCpuTargets = [ ];
```
Comment (alyssais) [inline pkgs/top-level/all-packages.nix:34939]: Shouldn't this be implied by `toolsOnly`?

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -34933,7 +34933,12 @@ with pkgs;
     inherit (darwin) sigtool;
   };
 
-  qemu-utils = callPackage ../applications/virtualization/qemu/utils.nix { };
+  qemu-utils = callPackage ../applications/virtualization/qemu/utils.nix {
```
Comment (alyssais) [inline pkgs/top-level/all-packages.nix:34936]: Why do we still need utils.nix?  How much space are we saving?

Comment (roberth) [reply pkgs/top-level/all-packages.nix:34936]: Some 35 MB on top of the 7 MB we had. It appears to be more useful and correct this way though.

File: pkgs/top-level/all-packages.nix (hunk 1 above)
Comment (roberth) [reply pkgs/top-level/all-packages.nix:34939]: Done.

File: pkgs/applications/virtualization/qemu/default.nix [hunk 3]
```diff
@@ -243,10 +248,62 @@ stdenv.mkDerivation (finalAttrs: {
     ln -s $out/bin/qemu-system-${stdenv.hostPlatform.qemuArch} $out/bin/qemu-kvm
   '';
 
+  # Normal qemu package does not have installcheck make target.
+  # toolsOnly has custom check.
+  doInstallCheck = toolsOnly;
+
+  installCheckPhase =
+    if toolsOnly
+    then ''
```
Comment (alyssais) [inline pkgs/applications/virtualization/qemu/default.nix:257]: lib.optionalString

File: pkgs/applications/virtualization/qemu/default.nix [hunk 4]
```diff
@@ -243,10 +248,62 @@ stdenv.mkDerivation (finalAttrs: {
...
+            qemu-img) ;;
+            qemu-io) ;;
+            qemu-nbd) ;;
+            qemu-pr-helper) ;;
+            qemu-storage-daemon) ;;
+            qemu-kvm)
+              if test -L $bin && ! test -e $bin; then
+                echo "removing broken $bin symlink"
+                echo "upstream build system should not write it, but it does, and it's not a big deal"
+                rm $bin
```
Comment (alyssais) [inline pkgs/applications/virtualization/qemu/default.nix:278]: This seems like a very weird thing to be doing in installCheckPhase.

File: pkgs/applications/virtualization/qemu/default.nix [hunk 5]
```diff
@@ -243,10 +248,62 @@ stdenv.mkDerivation (finalAttrs: {
...
+            qemu-edid) ;;
+            qemu-img) ;;
+            qemu-io) ;;
+            qemu-nbd) ;;
+            qemu-pr-helper) ;;
+            qemu-storage-daemon) ;;
+            qemu-kvm)
+              if test -L $bin && ! test -e $bin; then
+                echo "removing broken $bin symlink"
+                echo "upstream build system should not write it, but it does, and it's not a big deal"
```
Comment (alyssais) [inline pkgs/applications/virtualization/qemu/default.nix:277]: Can we fix this?

File: pkgs/applications/virtualization/qemu/default.nix [hunk 6]
```diff
@@ -243,10 +248,62 @@ stdenv.mkDerivation (finalAttrs: {
...
+        for bin in *; do
+          case $bin in
+            elf2dmp) ;;
+            qemu-edid) ;;
+            qemu-img) ;;
+            qemu-io) ;;
+            qemu-nbd) ;;
+            qemu-pr-helper) ;;
+            qemu-storage-daemon) ;;
+            qemu-kvm)
```
Comment (alyssais) [inline pkgs/applications/virtualization/qemu/default.nix:274]: qemu-kvm is not something the upstream build system generates.

File: pkgs/applications/virtualization/qemu/default.nix [hunk 7]
```diff
@@ -243,10 +248,62 @@ stdenv.mkDerivation (finalAttrs: {
...
+              else
+                echo "$bin is not a broken symlink, which is, oddly enough, unexpected"
+                echo "make sure the build only produced the tools and nothing else"
+              fi
+              ;;
+            qemu-system-*) # custom message, but just a special case of *)
+              echo "unexpected binary in qemu-utils package: $bin"
+              echo "qemu-utils is not supposed to contain the emulator"
+              exit 1
+              ;;
```
Comment (alyssais) [inline pkgs/applications/virtualization/qemu/default.nix:288]: This seems overly defensive to me.  We're explicitly telling it not to build any system emulators.  I don't think we need to second guess the build system.

**Review by alyssais — DISMISSED** [review]

File: pkgs/applications/virtualization/qemu/default.nix (hunk 3 above)
Comment (roberth) [reply pkgs/applications/virtualization/qemu/default.nix]: `null` avoided a rebuild of `qemu`, but I guess `optionalString` (`""`) is more idiomatic.

File: pkgs/applications/virtualization/qemu/default.nix (hunk 4 above)
Comment (roberth) [reply pkgs/applications/virtualization/qemu/default.nix]: I made a mistake. Thanks!

File: pkgs/applications/virtualization/qemu/default.nix (hunk 5 above)
Comment (roberth) [reply pkgs/applications/virtualization/qemu/default.nix]: Done.

File: pkgs/applications/virtualization/qemu/default.nix (hunk 7 above)
Comment (roberth) [reply pkgs/applications/virtualization/qemu/default.nix]: Removed the special case, but not the entire check. I think it's good to control which binaries are in the package.

Comment (alyssais) [reply pkgs/applications/virtualization/qemu/default.nix]: I'm not sure — given that we're happy including everything that's currently there, I think it's very likely we'll allow everything in the future anyway.

Comment (roberth) [reply pkgs/applications/virtualization/qemu/default.nix]: Ok. I've removed the check.

File: pkgs/top-level/all-packages.nix [hunk 8]
```diff
@@ -34933,7 +34933,9 @@ with pkgs;
     inherit (darwin) sigtool;
   };
 
-  qemu-utils = callPackage ../applications/virtualization/qemu/utils.nix { };
+  qemu-utils = qemu.override (o: {
```
Comment (alyssais) [inline pkgs/top-level/all-packages.nix:34936]: `override` can take an attrset — no need for a function.

**Review by alyssais — APPROVED** [review]
LGTM with a squash.

**Comment by alyssais** [comment]
BTW a future enhancement possibly worth considering would be _only_ building these tools in qemu-utils, and then symlinking to them from the other qemu packages, so we'd only build them once, and deduplicate them.

**Comment by roberth** [comment]
If the configuration flags don't affect the utils too much, that should work.
Even if some flags need to be propagated, the qemu-utils derivation might still converge to like two variations or something.

**Pull request merged**